internal/versioner/external.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package versioner
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func init() {
+	// Register the constructor for this type of versioner with the name "external"
+	Factories["external"] = NewExternal
+}
+
+// The type holds our configuration
+type External struct {
+	command    string
+	folderPath string
+}
+
+// The constructor function takes a map of parameters and creates the type.
+func NewExternal(folderID, folderPath string, params map[string]string) Versioner {
+	command := params["command"]
+
+	s := External{
+		command:    command,
+		folderPath: folderPath,
+	}
+
+	if debug {
+		l.Debugf("instantiated %#v", s)
+	}
+	return s
+}
+
+// Move away the named file to a version archive. If this function returns
+// nil, the named file does not exist any more (has been archived).
+func (v External) Archive(filePath string) error {
+	_, err := os.Lstat(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if debug {
+				l.Debugln("not archiving nonexistent file", filePath)
+			}
+			return nil
+		}
+		return err
+	}
+
+	if debug {
+		l.Debugln("archiving", filePath)
+	}
+
+	inFolderPath, err := filepath.Rel(v.folderPath, filePath)
+	if err != nil {
+		return err
+	}
+
+	cmd := exec.Command(v.command, v.folderPath, inFolderPath)
+	err = cmd.Run()
+	if err != nil {
+		l.Warnln("Versioner:", err)
```

Comment:
You need to do a `cmd.Wait()` as you'll end up with zombie processes.
Also, you should perhaps bail in the command is `""`, as you cannot guarantee that someone hasn't messed around with the config manually.

Also, now sure how much value (or rather concern) there is to this, but perhaps we should pass in an empty environment (rather than ours) to make sure we're not leaking any API keys or whatsoever just incase they were given to us via the environment.